### PR TITLE
Stop faking delegates not accessible to DynamicProxy

### LINKED ISSUE #1508: Stop faking delegates that are not visible to DynamicProxy in 5.0.0
Connects to #1437, in that if DynamicProxy gives us a way to fake delegates, we would likely have to drop this support anyhow. And it would make our capabilities more consistent, since we cannot fake classes or interfaces that aren't visible to DynamicProxy.

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fixes #1508

### CODE REVIEW

**Comment by blairconrad** [comment]
I don't mind it so much. I'd probably add a comment explaining why we're doing this craziness. Also, does it work well with generic delegates that act on inaccessible types, such as `Func<PrivateClass>`?

**Comment by thomaslevesque** [comment]
> I don't mind it so much. I'd probably add a comment explaining why we're doing this craziness.

Good point, will do.

> Also, does it work well with generic delegates that act on inaccessible types, such as `Func<PrivateClass>`?

I have no idea. I'll try.

**Comment by thomaslevesque** [comment]
> Also, does it work well with generic delegates that act on inaccessible types, such as `Func<PrivateClass>`?

Nope. Well, it works *too* well... it doesn't throw. Apparently `ProxyUtil.IsAccessible` doesn't look at the type parameters...

**Comment by thomaslevesque** [comment]
OK, it should work now

File: src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs [hunk 1]
```diff
@@ -1,28 +1,70 @@
...
+    using Castle.DynamicProxy;
     using FakeItEasy.Configuration;
     using FakeItEasy.Core;
 
     internal static class DelegateProxyGenerator
     {
+        private static readonly ProxyGenerator ProxyGenerator = new ProxyGenerator();
+        private static readonly ConcurrentDictionary<Type, bool> AccessibleToDynamicProxyCache = new ConcurrentDictionary<Type, bool>();
+
+
```
Comment (blairconrad) [inline src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs]: the thing that warns about extra blank lines warns about this extra blank line

File: tests/FakeItEasy.Specs/CreationSpecs.cs [hunk 2]
```diff
@@ -325,6 +325,44 @@ public void PrivateClassCannotBeFaked(Exception exception)
...
+
+            "When I create a fake of the delegate"
+                .x(() => exception = Record.Exception(this.CreateFake<Func<PrivateClass>>));
+
+            "Then it throws a fake creation exception"
+                .x(() => exception.Should().BeOfType<FakeCreationException>());
+
+            "And the exception message indicates the reason for failure"
+                .x(() => exception.Message.Should().Match(@"
+  Failed to create fake of type System.Func`1[FakeItEasy.Specs.CreationSpecsBase+PrivateClass]:
```
Comment (blairconrad) [inline tests/FakeItEasy.Specs/CreationSpecs.cs]: Ideally we'd be warning that we can't fake `Func<PrivateClass>`, as we're not actually trying to fake `PrivateClass`.
Sorry.

Comment (thomaslevesque) [reply tests/FakeItEasy.Specs/CreationSpecs.cs]: Isn't it what we're doing?

**Comment by blairconrad** [comment]
@thomaslevesque, webhooks are not reporting statuses or something, but you might as well squash it up, if you like. :+1:

**Comment by thomaslevesque** [comment]
Squashed!

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque!

**Pull request merged**